### PR TITLE
Feature custom manager

### PR DESCRIPTION
**Pull request by peterfarrell**
Add support for auto decryption of symmetric key and public key fields via the PGPManager (and support for disabling it in the Django Admin via the PGPAdmin). Includes unit tests and updated README.

This would close issue #5 if the PR is accepted.

### CODE REVIEW

File: pgcrypto/managers.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                elif isinstance(field, PGPPublicKeyFieldMixin):
+                    select_sql[field.name] = self._get_pgp_public_key_decrypt_sql(field)
+                    encrypted_fields.append(field.name)
+
+            # Django queryset.extra() is used here to add decryption sql to query.
+            qs = qs.defer(
+                *encrypted_fields
+            ).extra(
+                select=select_sql
+            )
```
Comment (KevinEtienne) [inline pgcrypto/managers.py]: Good idea! I wonder if it's possible to avoid to use `extra` but that might be for another PR. I suspect it will be not trivial, as the django field is not aware of its SQL implementation.

Comment (peterfarrell) [inline pgcrypto/managers.py]: Can't take all the credit -- saw something like this in another project and thought the same technique could be used here. I couldn't come up with something that was as straight forward as using `defer` and `extra`. I'm all ears for a better way to do this...

File: tests/test_fields.py [hunk 2]
```diff
@@ -408,3 +406,43 @@ def test_null(self):
...
+            hmac_field=expected_string,
+            pgp_pub_field=expected_string,
+            pgp_sym_field=expected_string,
+            date_pgp_sym_field=expected_date,  # Tests cast sql
+            datetime_pgp_sym_field=expected_datetime,  # Tests cast sql
+        )
+
+        instance = self.model.objects.get()
+
+        # Using `__dict__` bypasses the only the fly decryption for each field
```
Comment (KevinEtienne) [inline tests/test_fields.py]: What do you mean here?

File: tests/test_fields.py [hunk 3]
```diff
@@ -408,3 +406,43 @@ def test_null(self):
...
+        # If decryption is not working, we get references to <In_Memory> classes
+        self.assertEqual(instance.__dict__['pgp_pub_field'], expected_string)
```
Comment (KevinEtienne) [inline tests/test_fields.py]: what about?
```
self.assertEqual(instance.pgp_pub_field, expected_string)
```

File: README.md [hunk 4]
```diff
@@ -155,7 +161,36 @@ Example:
...
+If you use the bundled `PGPManager` with your custom model manager, all encrypted 
+fields will automatically decrypted for you (except for hash fields which are one 
+way).
+
+N.B. The bundled manager does not support decryption of fields from FK joins. For 
+example if the `MyModel` class had a FK to to `AnotherModel` class, no encrypted 
+fields be decrypted in the joined `AnotherModel`.
+
+It is recommended that you use the bundled `PGPAdmin` class if using the custom 
+model manager and the Django Admin. The Djano Admin performance suffers when
```
Comment (KevinEtienne) [inline README.md]: `:%s/The Djano/The Django`

File: tests/test_fields.py (hunk 3 above)
Comment (peterfarrell) [inline tests/test_fields.py]: Because if you switch the test to use `model = EncryptedModel` without the Manager, using the `instance.field_name` does on the fly decryption using another SQL query to decrypt that one field by itself. Using the __dict__ bypasses that feature in the project.

File: README.md (hunk 4 above)
Comment (peterfarrell) [inline README.md]: Fixed.

File: tests/test_fields.py (hunk 3 above)
Comment (KevinEtienne) [inline tests/test_fields.py]: Oh I see thanks.

File: tests/test_fields.py (hunk 2 above)
Comment (peterfarrell) [inline tests/test_fields.py]: See comment on the next review below...

File: tests/test_fields.py (hunk 3 above)
Comment (peterfarrell) [inline tests/test_fields.py]: Yea, otherwise the test isn't really testing the custom manager's decryption using `defer` and `extra`. Using `__dict__` is definitely a cheat to get the test to actually test what we were expected without changing how the project works.

File: tests/test_fields.py (hunk 2 above)
Comment (peterfarrell) [inline tests/test_fields.py]: Maybe you are referring to the `get()` here... There is only one row is the db so this is hack to get the last insert without knowing the pk.

> If you expect a queryset to return one row, you can use get() without any arguments to return the object for that row...

https://docs.djangoproject.com/en/1.10/ref/models/querysets/#django.db.models.query.QuerySet.get

Comment (KevinEtienne) [inline tests/test_fields.py]: I think as a non bative English speaker I'm having difficulty to understand this sentence construct `bypasses the only the fly`

Comment (peterfarrell) [inline tests/test_fields.py]: #notYourFault, it's a terribly written comment. Let me review that...

Comment (KevinEtienne) [inline tests/test_fields.py]: Cheers thanks.

**Comment by KevinEtienne** [comment]
Looks good to merge.

**Comment by peterfarrell** [comment]
Yay! Thanks! Do you know when you might be able to make a release on PyPI? https://pypi.python.org/pypi/django-pgcrypto-fields

**Comment by KevinEtienne** [comment]
I should be able to do it one now.
